src/main/scala/com/advancedtelematic/ota/deviceregistry/GroupMembership.scala

```
@@ -0,0 +1,98 @@
+package com.advancedtelematic.ota.deviceregistry
+
+import akka.http.scaladsl.util.FastFuture
+import com.advancedtelematic.libats.data.DataType.Namespace
+import com.advancedtelematic.libats.data.PaginationResult
+import com.advancedtelematic.ota.deviceregistry.common.Errors
+import com.advancedtelematic.ota.deviceregistry.data.Group.{GroupExpression, GroupId, Name}
+import com.advancedtelematic.ota.deviceregistry.data.GroupType.GroupType
+import com.advancedtelematic.ota.deviceregistry.data.{Group, GroupType, Uuid}
+import com.advancedtelematic.ota.deviceregistry.db.{DeviceRepository, GroupInfoRepository, GroupMemberRepository}
+import slick.jdbc.MySQLProfile.api._
+
+import scala.concurrent.{ExecutionContext, Future}
+
+protected trait GroupMembershipOperations {
```

IMO we do not need this trait. There is no case when we need it, `runGroupOperation` should throw an exception if group is a dynamic one

---

yes but then you have to check all the time if group is dynamic or not, this encapsulates that logic.

---

It does not, `runGroupOperation` does. And it does not need the trait.

---

Then you cannot throw different exceptions depending on the method unless you pass that exception arround, then the GroupMembership would have to know how to pass around different exceptions instead of just calling a method.

This is a tiny point but I think the trait makes it easier to read, but anyway this project is kind of a mess already.